#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

christophs78 opened the pull request:
See also https://github.com/primefaces/showcase-facelift/pull/41. The updated Showcase contains separated xhtml-pages and ViewBeans for DatePicker and Calendar with java.util.Date and java.time.LocalDate/LocalDateTime/LocalTime.

Backward-compatibility with java.util.Date is provided OOTB. So no migration work necessary. 
OOTB-backward-compatibility may be removed with a future (major) release.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

first iteration; work in progress

melloware commented:
Looks like a lot of work went into this.  Nice work.  I guess you need answers on all the TODO comments?

christophs78 commented:
Yes, it would help to get answers/opinions on the TODO comments.

tandraschko commented on src/main/java/org/primefaces/component/calendar/BaseCalendarRenderer.java:
```diff
@@ -144,12 +153,14 @@ public Object getConvertedValue(FacesContext context, UIComponent component, Obj
-                if (type != null && type != Object.class && type != Date.class) {
+                //if (type != null && type != Object.class && type != Date.class) {
+                if (type != null && type != Object.class && type != LocalDate.class && type != LocalDateTime.class && type != LocalTime.class) {
+                    //TODO: right decision to check Java8-Types instead of java.util.Date?
```
why not? :D

tandraschko commented on src/main/java/org/primefaces/component/api/UICalendar.java:
```diff
@@ -153,19 +155,49 @@ public boolean hasTime() {
             timeOnlyPattern = localePattern + " " + userTimePattern;
+             */
+
+            //TODO: Default-Pattern when no Pattern is specified via Pattern-Attribute?
```
shouldn't we just add a default pattern in the stateHelper call?

christophs78 replied:
Compared to java.util.Date - Code and think the easy solution is equivalent to the old behavior.

tandraschko commented on src/main/java/org/primefaces/component/api/UICalendar.java:
```diff
@@ -217,6 +249,7 @@ public void setLabelledBy(String labelledBy) {
     public Converter getConverter() {
         Converter converter = super.getConverter();
 
+        //TODO: Why does it matter whether Client-Side-Validation is enabled?
```
no idea, could you try to debug and/or remove this and check what happens?

christophs78 replied:
I tried both variant´s (via primefaces.CLIENT_SIDE_VALIDATION) but got no real idea. Thought about throwing this out.
Maybe there is some missalignement between the client-side-validation and the server-side-validation. And this code tries to work arround this issue. Hoped someone reminds why this was implemented this way some time ago.

christophs78 replied:
Checked how this was implemented in 6.2. (Before Datepicker)
https://github.com/primefaces/primefaces/blob/6_2/src/main/java/org/primefaces/component/calendar/CalendarRenderer.java
6.2 did not had this (from my perspective) awkward behaviour depending on Client-Side-Validation.
So i´ll throw this out. At least as long until someone come´s up with an argument, why this needs to be this way.
(This allows me to rollback org.primefaces.convert.DateTimeConverter to the original version.)

tandraschko commented on src/main/java/org/primefaces/component/calendar/BaseCalendarRenderer.java:
```diff
@@ -163,20 +174,49 @@ public Object getConvertedValue(FacesContext context, UIComponent component, Obj
-        format.setTimeZone(CalendarUtils.calculateTimeZone(uicalendar.getTimeZone()));
         try {
-            return format.parse(submittedValue);
+            //TODO: is there some way to dynamically determine the type?
```
which type?

christophs78 replied:
The type of the value which is bound to the component.
This would allow us to do something like the following pseudocode:
```
if (typeOfValue == java.util.Date) {
   //convert to java.util.Date
}
else if (typeOfValue == LocalDate)
{
   //convert to java.time.LocalDate
}
else if (typeOfValue == java.time.LocalDateTime)
{
   //convert to java.time.LocalDateTime
}
else if (typeOfValue == java.time.ZonedDateTime)
{
   //convert to java.time.ZonedDateTime
}
```
So we won´t need a converter for java.util.Date, we would be able to support ZonedDateTime.

christophs78 replied:
Maybe i simply have to look a few lines up.
```
ValueExpression ve = uicalendar.getValueExpression("value");
            if (ve != null) {
                Class type = ve.getType(context.getELContext());
```
I will try this.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

Improved Backward-Compatibility with java.util.Date.

christophs78 commented:
Backward-Compatibility with java.util.Date:
My first iteration required a converter.
`<p:datePicker id="inline" value="#{calendarView.date1}" inline="true" converter="pf.DateBackwardCompatiblityConverter" />`

My second iteration determines the type of the value expression dynamical. So no migration work necessary. Only exception is DatePicker with selectionMode = multiple and selectionMode = range.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

Improved Backward-Compatibility with java.util.Date.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

java.util.TimeZone --> java.time.ZoneId

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

BaseCalendarRenderer fix FacesMessage for unsupported types

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

Remove awkward switch to DateTimeConverter when primefaces.CLIENT_SIDE_VALIDATIONS is enabled.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

work on TODOs

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

work on TODOs

tandraschko commented:
AFAICS if we can determine the generic type of the list, we could also remove DateBackwardCompatiblityConverter?

christophs78 commented:
Yes.

tandraschko commented:
Ok, just google a bit. This should be possible via TypeArgument API.
If not, let me know

christophs78 commented:
I already googled a few days ago...
https://stackoverflow.com/questions/1942644/get-generic-type-of-java-util-list
At runtime there is no typing Information. Only the compiler knows how the list is typed. At runtime it´s a untyped list.
The only way to find out the type  of the list at runtime would be reflections. But I do not think this works together with value-binding via EL.

tandraschko commented:
Hmmm, not "easy" but it's possible:
Get the real value expression / reference via: ValueExpressionAnalyzer.getReference and get the field via reflection. Then you can also get the type as you already mentioned.

This way it will also work with a null list.

christophs78 commented:
Sounds good. I will try this tomorrow. :-)

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

remove DateBackwardCompatiblityConverter
BaseCalendar Renderer - dynamically determine type of java.util.list

christophs78 commented:
It works! :-)
Hope we are able to merge this PR within the next days.
After this is done i´ll work on a second PR which adds LocalDate/Time-Support to Schedule and Timeline.

tandraschko commented:
great :)
I will try to do a big review on sunday

christophs78 commented:
Is there a chance to make some progress on this PR? I´d realy like to continue with Timeline, Schedule, ... after this is done. :-)

tandraschko commented:
Will review next month, sry

tandraschko commented:
@christophs78 the changes are looking fine! I assume that i you tested everything (and also CSV)?

christophs78 commented:
@tandraschko: Thank you for reviewing! I tested everything that came into my mind. (As you can see in the PR in the showcase-project, quite a good amount of testcases.)
CSV??? Do you mean comma separated values? Hm. During CSV-export of a DataTable which contains Calendar or Datepicker? Please help me. :-)

melloware commented:
No CSV = Client Side Validation.   https://www.primefaces.org/showcase/ui/csv/basic.xhtml

christophs78 commented:
https://github.com/primefaces/primefaces/blob/4c435187cf75befa2a53e4b923198fe9359531ba/src/main/resources/META-INF/resources/primefaces/validation/validation.js#L435

First assessment:
On the client-side it should not matter, which type we have on the server. Relevant is the data-p-con-Attribute:
`<input id="form:popup_input" ... data-p-con="javax.faces.DateTime"`

Seems CSV for Calendar/DatePicker required to explicit use <f:convertDateTime pattern="xxx" />. This does not change.

I´lll create a testcase to be totally sure.

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

ClientSideValidation

christophs78 commented:
CSV is fixed. :-)
Thanks for drawing my attention to this!

NicolaIsotta commented:
@christophs78 would it be possible to add `java.time.YearMonth` too? I think it combines well with `<p:datePicker view="month"/>`

christophs78 committed:
#4646 Move to Java8 LocalDate/Time (Calendar + Datepicker)

java.time.YearMonth

christophs78 commented:
@NicolaIsotta: Done :-)
